server/client-connection.cpp

```diff
@@ -242,9 +243,46 @@ void ClientConnection::process_message(MessageIn& msg) {
 		EventSystem<ChatCommandEvent>::Get()->Emit(std::make_shared<ChatCommandEvent>(chat_command));
 		break;
 	}
+	case MessageType::LOGIN:
+	{
+		proto::UserLogin user_login;
+		user_login.ParseFromZeroCopyStream(&msg);
+		UserLoginInfo user_login_event{user_login.username()};
+
+		auto _user = this->server->GetAuthenticator().Authenticate(user_login.username());
+		user_login_event.authenticated = _user != nullptr;
+		user_login_event.user_id = _user ? _user->GetUserId() : "";
+
+		this->server->GetLuaSystem()->CallFunctions("onUserLogin", &user_login_event);
+
+		if (!user_login_event.reject && user_login_event.authenticated) {
+			this->user = _user; // Could be null, if authenticated was overriden in script.
+
+			auto authd_message = MessagePool::get();
+			authd_message->SetBodyLength(1);
+			authd_message->SetMessageType(MessageType::AUTHENTICATED);
+			authd_message->encode_header();
+			QueueWrite(authd_message);
+
+			this->server->SendWorld(shared_from_this());
+		}
+		else {
+			this->server->OnDisconnect(shared_from_this());
```

Comment:
There isn't a reject message yet and I didn't want to add even more to this PR. Plus I think coupling it will a true kick mechanic might be a good option